presto-hive/src/main/java/io/prestosql/plugin/hive/s3/PrestoS3FileSystem.java

```diff
@@ -1400,6 +1462,195 @@ public synchronized void progressChanged(ProgressEvent progressEvent)
+
+            this.parts = new ArrayList<>();
+        }
+
+        @Override
+        public void write(int b)
+                throws IOException
+        {
+            if (count >= buf.length) {
+                flushBuffer(false);
+            }
+
+            buf[count] = (byte) b;
+            count++;
+        }
+
+        @Override
+        public void write(byte[] bytes, int off, int len)
+                throws IOException
+        {
+            int offset = off;
+            int remain = len;
+            while ((buf.length - count) <= remain) {
+                int copied = buf.length - count;
+                arraycopy(bytes, offset, buf, count, copied);
+                count += copied;
+
+                flushBuffer(false);
+
+                offset += copied;
+                remain -= copied;
+            }
+
+            if (remain > 0) {
+                arraycopy(bytes, offset, buf, count, remain);
+                count += remain;
+            }
+        }
+
+        @Override
+        public void flush()
+                throws IOException
+        {
+            flushBuffer(false);
+        }
+
+        @Override
+        public void close()
+                throws IOException
+        {
+            try {
+                flushBuffer(true);
+                waitForPreviousUploadFinish();
+            }
+            catch (IOException e) {
+                abortUpload(true);
+                throw e;
+            }
+
+            if (!parts.isEmpty()) {
+                try {
+                    finishUpload(parts);
+                }
+                catch (Exception e) {
+                    abortUpload(true);
+                    throw new IOException(e);
+                }
+            }
+            else {
+                abortUpload(false);
+            }
+        }
+
+        private void flushBuffer(boolean finished)
+                throws IOException
+        {
+            // The multipart upload API only accept the last part to be less than 5MB
+            if (count == buf.length || (finished && count > 0)) {
+                byte[] data = buf;
+                int length = count;
+                this.buf = new byte[buf.length];
+                count = 0;
+
+                waitForPreviousUploadFinish();
+
+                inProgressUploadFuture = uploadExecutor.submit(() -> uploadPage(data, length));
+            }
+        }
+
+        private void waitForPreviousUploadFinish()
+                throws IOException
+        {
+            if (inProgressUploadFuture == null) {
+                return;
+            }
+
+            try {
+                inProgressUploadFuture.get();
+            }
+            catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                throw new InterruptedIOException();
+            }
+            catch (ExecutionException e) {
+                throw new IOException("Streaming upload failed", e.getCause());
```

Comment:
It's better to throw `e` here so that we retain the full stack trace. While the `ExecutionException` is not really relevant to the underlying exception, it contains the `Future.get()` call in the stack. If you discard it, you see a stack trace from a different thread, which can be confusing.

See https://github.com/google/guava/wiki/Why-we-deprecated-Throwables.propagate#encourages-unwrapping-exceptions-from-other-threads which changed my opinion on this